File: game.py
import pygame
import random
import cv2
import mediapipe as mp
import time

# ========== Gesture Setup ==========
cap = cv2.VideoCapture(0)
mp_hands = mp.solutions.hands
hands = mp_hands.Hands(max_num_hands=1, min_detection_confidence=0.7)
jump_gesture_cooldown = 0
jump_delay = 1.0

# ========== Game Setup ==========
pygame.init()
WIDTH, HEIGHT = 800, 400
win = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Dino Jump with Gesture")

clock = pygame.time.Clock()
font = pygame.font.SysFont("Arial", 30)

# Colors
WHITE = (255,255,255)
BLACK = (0,0,0)
RED = (200,0,0)
GREEN = (0,200,0)

# Dino
dino = pygame.Rect(100, 300, 50, 50)
gravity = 0.6
velocity = 0
jumping = False

# Cactus
cactus = pygame.Rect(WIDTH, 310, 20, 40)
cactus_speed = 6

# Score
score = 0
game_over = False

# Gesture detection
def is_jump_gesture(landmarks):
    if not landmarks:
        return False
    fingers = []
    tips = [8, 12, 16, 20]
    fingers.append(1 if landmarks[4].x < landmarks[3].x else 0)  # Thumb
    for tip in tips:
        fingers.append(1 if landmarks[tip].y < landmarks[tip - 2].y else 0)
    return sum(fingers) == 5  # Full hand open

# ========== Game Loop ==========
running = True
while running:
    win.fill(WHITE)

    # Check gesture input
    ret, frame = cap.read()
    if not ret: break
    frame = cv2.flip(frame, 1)
    rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    results = hands.process(rgb)
    now = time.time()

    if results.multi_hand_landmarks:
        lmList = results.multi_hand_landmarks[0].landmark
        if is_jump_gesture(lmList) and not jumping and now - jump_gesture_cooldown > jump_delay:
            velocity = -12
            jumping = True
            jump_gesture_cooldown = now

    # Dino jump physics
    velocity += gravity
    dino.y += velocity
    if dino.y >= 300:
        dino.y = 300
        velocity = 0
        jumping = False

    # Move cactus
    cactus.x -= cactus_speed
    if cactus.x < -20:
        cactus.x = WIDTH + random.randint(0, 200)
        score += 1

    # Collision detection
    if dino.colliderect(cactus):
        game_over = True

    # Draw
    pygame.draw.rect(win, GREEN, dino)
    pygame.draw.rect(win, RED, cactus)
    score_text = font.render(f"Score: {score}", True, BLACK)
    win.blit(score_text, (10, 10))

    if game_over:
        over_text = font.render("Game Over! Press R to Restart", True, RED)
        win.blit(over_text, (WIDTH//2 - 180, HEIGHT//2))
        pygame.display.update()
        key = None
        while key != pygame.K_r:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    running = False
                    key = pygame.K_r
                elif event.type == pygame.KEYDOWN:
                    key = event.key
        # Reset
        dino.y = 300
        cactus.x = WIDTH
        velocity = 0
        jumping = False
        score = 0
        game_over = False

    pygame.display.update()
    clock.tick(60)

    # Exit key
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

cap.release()
pygame.quit()
cv2.destroyAllWindows()

File: game.js
let dino = { y: 300, vy: 0, gravity: 1.5, jumping: false };
let canvas = document.getElementById('gameCanvas');
let ctx = canvas.getContext('2d');
let gestureStatus = document.getElementById('gestureStatus');

function startGame() {
  gestureStatus.innerText = 'Gesture: Waiting...';
  setupCameraAndGesture();
  requestAnimationFrame(update);
}

function jump() {
  if (!dino.jumping) {
    dino.vy = -20;
    dino.jumping = true;
  }
}

function update() {
  ctx.clearRect(0, 0, canvas.width, canvas.height);

  // gravity
  dino.y += dino.vy;
  dino.vy += dino.gravity;

  if (dino.y >= 300) {
    dino.y = 300;
    dino.vy = 0;
    dino.jumping = false;
  }

  ctx.fillStyle = 'brown';
  ctx.fillRect(100, dino.y, 50, 50);

  requestAnimationFrame(update);
}

function setupCameraAndGesture() {
  const video = document.getElementById('video');
  const hands = new Hands({
    locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`
  });

  hands.setOptions({
    maxNumHands: 1,
    modelComplexity: 0,
    minDetectionConfidence: 0.7,
    minTrackingConfidence: 0.7
  });

  hands.onResults((results) => {
    if (results.multiHandLandmarks.length > 0) {
      gestureStatus.innerText = 'Gesture: Hand Detected';
      const landmarks = results.multiHandLandmarks[0];

      // basic fist check
      const isFist = landmarks[8].y > landmarks[6].y && landmarks[12].y > landmarks[10].y;
      if (isFist) {
        jump();
        gestureStatus.innerText = 'Gesture: FIST (Jump)';
      }
    } else {
      gestureStatus.innerText = 'Gesture: NONE';
    }
  });

  const camera = new Camera(video, {
    onFrame: async () => {
      await hands.send({ image: video });
    },
    width: 640,
    height: 480
  });

  camera.start();
}
